File: JAVA/lld-examples/easy/parking-lot-system/src/main/java/com/systemdesign/lld/parking/ParkingSpot.java
package com.systemdesign.lld.parking;

/**
 * Parking spot class
 */
public class ParkingSpot {
    private int spotNumber;
    private int floorNumber;
    private VehicleType vehicleType;
    private boolean isOccupied;
    private Vehicle parkedVehicle;
    private double hourlyRate;
    
    public ParkingSpot(int spotNumber, int floorNumber, VehicleType vehicleType) {
        this.spotNumber = spotNumber;
        this.floorNumber = floorNumber;
        this.vehicleType = vehicleType;
        this.isOccupied = false;
        this.parkedVehicle = null;
        this.hourlyRate = vehicleType.getHourlyRate();
    }
    
    public boolean isAvailable() {
        return !isOccupied;
    }
    
    public boolean canParkVehicle(Vehicle vehicle) {
        return isAvailable() && vehicle.getType() == this.vehicleType;
    }
    
    public boolean parkVehicle(Vehicle vehicle) {
        if (canParkVehicle(vehicle)) {
            this.isOccupied = true;
            this.parkedVehicle = vehicle;
            return true;
        }
        return false;
    }
    
    public void removeVehicle() {
        this.isOccupied = false;
        this.parkedVehicle = null;
    }
    
    // Getters
    public int getSpotNumber() {
        return spotNumber;
    }
    
    public int getFloorNumber() {
        return floorNumber;
    }
    
    public VehicleType getVehicleType() {
        return vehicleType;
    }
    
    public boolean isOccupied() {
        return isOccupied;
    }
    
    public Vehicle getParkedVehicle() {
        return parkedVehicle;
    }
    
    public double getHourlyRate() {
        return hourlyRate;
    }
    
    @Override
    public String toString() {
        return "Spot{" +
                "floor=" + floorNumber +
                ", spot=" + spotNumber +
                ", type=" + vehicleType +
                ", occupied=" + isOccupied +
                '}';
    }
} 

File: Python/lld-examples/easy/parking-lot-system/parking_spot.py
from typing import Optional
from vehicle import Vehicle, VehicleType

class ParkingSpot:
    """Parking spot class"""
    
    def __init__(self, spot_number: int, floor_number: int, vehicle_type: VehicleType):
        self.spot_number = spot_number
        self.floor_number = floor_number
        self.vehicle_type = vehicle_type
        self.is_occupied = False
        self.parked_vehicle: Optional[Vehicle] = None
        self.hourly_rate = vehicle_type.hourly_rate
    
    def is_available(self) -> bool:
        """Check if spot is available"""
        return not self.is_occupied
    
    def can_park_vehicle(self, vehicle: Vehicle) -> bool:
        """Check if vehicle can park in this spot"""
        return self.is_available() and vehicle.type == self.vehicle_type
    
    def park_vehicle(self, vehicle: Vehicle) -> bool:
        """Park vehicle in this spot"""
        if self.can_park_vehicle(vehicle):
            self.is_occupied = True
            self.parked_vehicle = vehicle
            return True
        return False
    
    def remove_vehicle(self):
        """Remove vehicle from spot"""
        self.is_occupied = False
        self.parked_vehicle = None
    
    def __str__(self):
        return f"Spot{{floor={self.floor_number}, spot={self.spot_number}, type={self.vehicle_type}, occupied={self.is_occupied}}}" 